Add a compute shader bonus chapter

SaschaWillems opened the pull request:
Note: I'm still working on this, but will probably mostly change or add minor things, so this is open for feedback :)

This PR adds a new bonus chapter about using compute shaders with Vulkan. It's at least partially based on [on this proposal](https://docs.google.com/document/d/1YLHRxS6UW6BL5UuopMq8nMQ1GHhdlO8b-xn_R_OJXr4) that we talked about in earlier calls.

I've been working on this for quite some time now and this is my fifth or sixth try and with which I'm mostly happy right now. I tried to find the right balance between an advanced topic like compute and the Vulkan Tutorial being aimed beginners. I also did some images to help people understand compute, e.g. how compute space is structured and why we use SSBO's per frame and why frame n+1 needs the SSBO from frame n as an input. So to sum it up: I tried to keep in the style and on the technical level of the other parts of the tutorial.

So along with writing the chapter I also added a new sample that renders a GPU based particle system. The chapter then refers to this code in a lot of places to (hopefully) make it easy for people to follow along both in written form as well as in code.

I also hope that the chapter includes all relevant compute information and in an order that beginners can easily follow.

The [markdown version can be found here](https://github.com/SaschaWillems/VulkanTutorial/blob/compute_bonus_chapter/en/11_Compute_Shader.md).

Areas that needs to be checked:
- Is the information complete (any important compute concept/term missing)?
- Is the chapter easy to follow?
- Are the explanations correct?
- Do the images make sense?
- Is the spelling okay (I'm not a native speaker)
- Is the synchronization correct?

Esp. the later I'm not 100% sure about. Both normal and sync validation are clean, so in terms of API usage sync is fine, but I may have overdone it a bit.

Opens:
- I haven't added the CMake changes that include the compute shader in the build files yet, I need to check first if this doesn't brake the other samples

SaschaWillems committed:
Started working on compute shader bonus chapter

SaschaWillems committed:
Continued work on compute shader bonus chapter

SaschaWillems committed:
Continued work on compute shader bonus chapter

SaschaWillems committed:
Added code for compute sample

work-in-progress

SaschaWillems committed:
Added code for compute sample

Per-frame ssbos
Compute shader is now integrated and updating particles
work-in-progress

SaschaWillems committed:
Added code for compute sample

Nicer visuals, circular initial particle positions
work-in-progress

SaschaWillems committed:
Added code for compute sample

work-in-progress

SaschaWillems committed:
Tutorial compute queue families, submission and synchronization chapter

SaschaWillems committed:
Added paragraphs on sync, submit and queue familes

Rewording some paragraphs

SaschaWillems committed:
Rewording

Added image for workgroup and invocations compute space

SaschaWillems committed:
Sample for work group and dispatch counts

SaschaWillems committed:
Note and spec reference on queue families

SaschaWillems committed:
Minor additions

SaschaWillems committed:
Sample code uses SSBO per frame

Code cleanup
Tutotorial changed

SaschaWillems committed:
Added image for SSBOs per-frame reads and writes

SaschaWillems committed:
Conclusion

charles-lunarg commented:
I like it!

Some thoughts:
* Is this order of topic introduction the best way? Not saying its wrong, but I was wondering if there were ways to make the chapter 'flow' better. Like, the discussion of workgroups & sizes happened a lot later than I was expecting. Plus, it was hard to tell which code snippets were explanatory and which were 'copy and paste this into your code' 
* What happens if you want to render a non multiple of 256 particles?
* Wont the fence need to have the first index already signaled? I think a quick reminder would be helpful.
* It seems like these particles are fixed size and never get created and never deleted. It would be cool to talk about 'adding & removing' them (but this might get out of scope pretty fast, since that gets very implementation detail heavy whereas we want to focus on the 'vulkan' aspects of compute shaders)
*

SaschaWillems commented:
Thank you very much for your feedback. Those are some good and valid points :)

> Is this order of topic introduction the best way? Not saying its wrong, but I was wondering if there were ways to make the chapter 'flow' better. Like, the discussion of workgroups & sizes happened a lot later than I was expecting. Plus, it was hard to tell which code snippets were explanatory and which were 'copy and paste this into your code'

Yeah, I'm not 100% happy with the order of topics. On one hand it's better to explain things before they are mentioned, on the other hand if you mention them too early (before they are used) readers might already have forgotten about them. I'll take a look at how to improve ordering a bit.

> What happens if you want to render a non multiple of 256 particles?

Right. I should explain that. Or in general how to e.g. work with different multiples.

> Wont the fence need to have the first index already signaled? I think a quick reminder would be helpful.

Correct, and that's what the code already does (and did for the graphics fence). That's why it's not mentioned yet. But as you said, it's an important info that I'll add.

> It seems like these particles are fixed size and never get created and never deleted. It would be cool to talk about 'adding & removing' them (but this might get out of scope pretty fast, since that gets very implementation detail heavy whereas we want to focus on the 'vulkan' aspects of compute shaders)

 While that would make sense in this chapter, I also think this could be a bit too much for the people that the tutorial is aimed at. At least for this chapter. Maybe we can add that to another chapter on rendering more dynamic stuff.

SaschaWillems committed:
Added note on signaled fences

marty-johnson59 committed:
Update 11_Compute_Shader.md

marty-johnson59 commented:
I did an editing pass through this to find/fix any typos/grammar issues.  I just did a commit directly to the branch - hopefully that doesn't break things...

SaschaWillems commented:
Thanks a lot. Just checked and yes, this works fine, all your changes are also visible in my fork repository :)

SaschaWillems committed:
Merge branch 'master' into compute_bonus_chapter

SaschaWillems committed:
Use persistent mapped buffers

charles-lunarg commented:
It seems that the changes to add persistently mapped uniform buffers are present here. That appears to be a mistake, and just need some rebasing to clean up the commit list.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+## Advantages
+
+Doing computationally expensive calculations on the GPU has several advantages. The most obvious one is offloading work from the CPU. Another one is not requiring a round-trip to the CPU and it's main memory, so all data can stay on the GPU without having to resolve to slow main memory reads.
```
```suggestion
Doing computationally expensive calculations on the GPU has several advantages. The most obvious one is offloading work from the CPU. Another one is not requiring moving data between the CPU's main memory and the GPU's memory. All of the data can stay on the GPU without having to wait for slow transfers from main memory.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+Doing computationally expensive calculations on the GPU has several advantages. The most obvious one is offloading work from the CPU. Another one is not requiring a round-trip to the CPU and it's main memory, so all data can stay on the GPU without having to resolve to slow main memory reads.
+
+Aside from these, GPUs are heavily parallelized with some of them having tens of thousands of small compute units. This often makes them a better fit for highly parallel workflows than a CPU with a few large units.
```
```suggestion
Aside from these, GPUs are heavily parallelized with some of them having tens of thousands of small compute units. This often makes them a better fit for highly parallel workflows than a CPU with a few large compute units.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+## The Vulkan pipeline
+
+It's important to know that compute is completely separated from the graphics part of the pipeline. This is visible in the following block diagram of the Vulkan pipeline from the official specification:
```
"The Vulkan Pipeline" is a concept that I am note sure is one readers are familiar with. The introduction chapter and graphics pipeline chapter all refer to the graphics pipeline, and do not refer to a concept that doesn't also include graphics somehow.

SaschaWillems replied:
Good point. Maybe a different caption like "Where does compute fit into compute?" would be better here?

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+## An example
+
+An easy to understand example that we implement in this chapter is a GPU based particle system. Such systems are used in many games and often consist of thousands of particles that need to be updated at interactive frame rates. And sometimes even with complex physics applied, e.g. when testing for collisions. So rendering such a system requires vertices (passed as vertex buffers) and a way to upd ...
```
```suggestion
An easy to understand example that we will implement in this chapter is a GPU based particle system. Such systems are used in many games and often consist of thousands of particles that need to be updated at interactive frame rates. Rendering such a system requires 2 main components: vertices, passed as vertex buffers, and a way to update them based on some equation.
```
(Dropping the sentence since it doesn't add much and greatly interrupts the flow)

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+An easy to understand example that we implement in this chapter is a GPU based particle system. Such systems are used in many games and often consist of thousands of particles that need to be updated at interactive frame rates. And sometimes even with complex physics applied, e.g. when testing for collisions. So rendering such a system requires vertices (passed as vertex buffers) and a way to upd ...
+
+A "classical" CPU based particle system would store particles in the system's main memory and then use the CPU to update them. After the update, the vertices need to be transferred to the GPU's memory again so it will display the updated particles in the next frame. The most straight-forward way would be recreating the vertex buffer with the new data for each frame. This is obviously very costly. ...
```
```suggestion
The "classical" CPU based particle system would store particle data in the system's main memory and then use the CPU to update them. After the update, the vertices need to be transferred to the GPU's memory again so it can display the updated particles in the next frame. The most straight-forward way would be recreating the vertex buffer with the new data for each frame. This is obviously very costly. Depending on your implementation, there are other options like mapping GPU memory so it can be written by the CPU (called "resizable BAR" on desktop systems, or unified memory on integrated GPUs) or just using a host local buffer (which would be the slowest method due to PCI-E bandwidth). But no matter what buffer update method you choose, you always require a "round-trip" to the CPU to update the particles.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+A "classical" CPU based particle system would store particles in the system's main memory and then use the CPU to update them. After the update, the vertices need to be transferred to the GPU's memory again so it will display the updated particles in the next frame. The most straight-forward way would be recreating the vertex buffer with the new data for each frame. This is obviously very costly. ...
+
+With a GPU based particle system, this round-trip is no longer required. Vertices are only uploaded to the GPU once and all updates are done in the GPU's memory by using compute shaders. One of the main reasons why this is faster is the much higher bandwidth between the GPU and it's local memory. In a CPU based scenario, you'd be limited by main memory and PCI-express bandwidth, which is often ju ...
```
```suggestion
With a GPU based particle system, this round-trip is no longer required. Vertices are only uploaded to the GPU at the start and all updates are done in the GPU's memory using compute shaders. One of the main reasons why this is faster is the much higher bandwidth between the GPU and it's local memory. In a CPU based scenario, you'd be limited by main memory and PCI-express bandwidth, which is often just a fraction of the GPU's memory bandwidth.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+### Shader storage buffer objects (SSBO)
+
+A shader storage buffer (SSBO) allows you to read from and write to a buffer. Using these is similar to using uniform buffer objects. The biggest differences are that you can alias other buffer types to SSBOs and that they can be arbitrarily large.
```
```suggestion
A shader storage buffer (SSBO) allows shaders to read from and write to a buffer. Using these is similar to using uniform buffer objects. The biggest differences are that you can alias other buffer types to SSBOs and that they can be arbitrarily large.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+...
+
+void createShaderStorageBuffers() {
+    ...
```
This is a new function, there is no need for an ellipsis (...) to indicate adding to other code.

SaschaWillems replied:
Yeah, but the actual sample code has some lines of code before what's shown here. hat's why I put the ellipsis in here.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+    for (size_t i = 0; i < MAX_FRAMES_IN_FLIGHT; i++) {
+        createBuffer(bufferSize, VK_BUFFER_USAGE_STORAGE_BUFFER_BIT | VK_BUFFER_USAGE_VERTEX_BUFFER_BIT | VK_BUFFER_USAGE_TRANSFER_DST_BIT, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT, shaderStorageBuffers[i], shaderStorageBuffersMemory[i]);
+        copyBuffer(stagingBuffer, shaderStorageBuffers[i], bufferSize);
```
Where does stagingBuffer get defined? More importantly, does it have enough space to accomodate the upload? Also this implies there is 'starting' data to upload which hasn't been discussed at this point in the chapter.

SaschaWillems replied:
That's true. I've added the relevant parts from the sample code so people can easily follow what happens. Also added some code comments.

charles-lunarg commented on en/11_Compute_Shader.md:
````diff
@@ -0,0 +1,601 @@
+```
+The two flags `VK_BUFFER_USAGE_VERTEX_BUFFER_BIT` and `VK_BUFFER_USAGE_STORAGE_BUFFER_BIT` set with `bufferInfo.usage` tell the implementation that we want to use this buffer for two different scenarios: as a vertex buffer in the vertex shader and as a store buffer. Note that we also added the `VK_BUFFER_USAGE_TRANSFER_DST_BIT` flag in here so we can transfer data from the host to the GPU. This i ...
+
+Here is the same code using using the `createBuffer` helper function:
````
I have a feeling introducing too much code at this stage in the tutorial is counter productive. The current section is "the two ways to read/write data in compute shaders: SSBO's and Images". Thus, the bits describing `createBuffer`, adding variables to hold the VkBuffer & VkDeviceMemory in accordance to frames in flight, and the actual initialization function should happen later on.

SaschaWillems replied:
That's a good point. I'll try to move the code heavy parts down the line so they're closer to the other parts of the actual tutorial.

SaschaWillems replied:
I have restructured this a bit. I moved most of the relevant code further down right before how descriptors are setup, so that this chapter at the beginning isn't too code heavy and similar to the storage image one.

charles-lunarg commented on en/11_Compute_Shader.md:
````diff
@@ -0,0 +1,601 @@
+}
+```
+
+The GLSL shader declaration for accessing such a buffer looks like this:
+
+```glsl
````
The glsl code here is fine however, since it highlights 'how to use ssbo's in shaders.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+A storage image allows you read from and write to an image. Typical use cases are applying image effects to textures or doing post processing (which in turn is very similar).
+
+This is similar for images:
```
This refers to what here?

charles-lunarg commented on en/11_Compute_Shader.md:
````diff
@@ -0,0 +1,601 @@
+}
+```
+
+The two flags `VK_IMAGE_USAGE_SAMPLED_BIT` and `VK_IMAGE_USAGE_STORAGE_BIT` set with `bufferInfo.usage` tell the implementation that we want to use this image for two different scenarios: as an image sampled in the fragment shader and as a storage image in the computer shader;
````
```suggestion
The two flags `VK_IMAGE_USAGE_SAMPLED_BIT` and `VK_IMAGE_USAGE_STORAGE_BIT` set with `imageInfo.usage` tell the implementation that we want to use this image for two different scenarios: as an image sampled in the fragment shader and as a storage image in the computer shader;
```

charles-lunarg commented on en/11_Compute_Shader.md:
````diff
@@ -0,0 +1,601 @@
+layout (binding = 1, rgba8) uniform writeonly image2D outputImage;
+```
+
+A few differences here are additional attributes like `rgba8` for the format of the image, the `readonly` and `writeonly` qualifiers, telling the implementation that we'll only read from the input and write to the output image. And last but not least we need to use the `image2D` type do declare a storage image.
````
```suggestion
A few differences here are additional attributes like `rgba8` for the format of the image, the `readonly` and `writeonly` qualifiers, telling the implementation that we will only read from the input image and write to the output image. And last but not least we need to use the `image2D` type to declare a storage image.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+## Compute queue families
+
+In the [physical device and queue families chapter](03_Drawing_a_triangle/00_Setup/03_Physical_devices_and_queue_families.md#page_Queue-families) we already learned about queue families and how to select a graphics queue family. Compute introduces the new queue family type bit `VK_QUEUE_COMPUTE_BIT`. So if we want to do compute, we need to get a queue from a queue family type that supports comput ...
```
```suggestion
In the [physical device and queue families chapter](03_Drawing_a_triangle/00_Setup/03_Physical_devices_and_queue_families.md#page_Queue-families) we already learned about queue families and how to select a graphics queue family. Compute uses the queue family properties flag bit `VK_QUEUE_COMPUTE_BIT`. So if we want to do compute work, we need to get a queue from a queue family that supports compute.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+In the [physical device and queue families chapter](03_Drawing_a_triangle/00_Setup/03_Physical_devices_and_queue_families.md#page_Queue-families) we already learned about queue families and how to select a graphics queue family. Compute introduces the new queue family type bit `VK_QUEUE_COMPUTE_BIT`. So if we want to do compute, we need to get a queue from a queue family type that supports comput ...
+
+Note that Vulkan requires an implementation that supports graphics operations to have at least one queue family that supports both graphics and compute, but it's also possible that implementations offer a dedicated compute queue. This dedicated compute queue (that does not have the graphics bit) hints at an asynchronous compute queue. To keep this tutorial beginner friendly though, we'll use a qu ...
```
```suggestion
Note that Vulkan requires an implementation which supports graphics operations to have at least one queue family that supports both graphics and compute operations, but it's also possible that implementations offer a dedicated compute queue. This dedicated compute queue (that does not have the graphics bit) hints at an asynchronous compute queue. To keep this tutorial beginner friendly though, we'll use a queue that can do both graphics and compute operations. This will also save us from dealing with several advanced synchronization mechanisms.
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+Note that Vulkan requires an implementation that supports graphics operations to have at least one queue family that supports both graphics and compute, but it's also possible that implementations offer a dedicated compute queue. This dedicated compute queue (that does not have the graphics bit) hints at an asynchronous compute queue. To keep this tutorial beginner friendly though, we'll use a qu ...
+
+So for our compute sample we need to change the device creation code a bit:
```
```suggestion
For our compute sample we need to change the device creation code a bit:
```

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+int i = 0;
+for (const auto& queueFamily : queueFamilies) {
+    if ((queueFamily.queueFlags & VK_QUEUE_GRAPHICS_BIT) && (queueFamily.queueFlags & VK_QUEUE_COMPUTE_BIT)) {
+        indices.graphicsAndComputeFamily = i;
```
Maybe show changing the struct member name as well? Also maybe should the queue be renamed to graphicsAndCompute?

SaschaWillems replied:
Not sure on this. There is already a graphicsQueue in there, so I thought it would be better to just name it computeQueue so people don't get too confused?

charles-lunarg commented on en/11_Compute_Shader.md:
````diff
@@ -0,0 +1,601 @@
+We can then get a compute queue from this queue family in `createLogicalDevice`:
+
+```c++
+vkGetDeviceQueue(device, indices.graphicsAndComputeFamily.value(), 0, &computeQueue);
````
Same as above, may need to change the name here.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,601 @@
+
+## Compute pipelines
+
+As compute is not a part of the graphics pipeline, we can't use `vkCreateGraphicsPipelines` to attach the compute shader to it. Instead we need to create a dedicated compute pipeline `vkCreateComputePipelines` for running our compute commands. Since a compute pipeline does not touch any of the rasterization state, it has a lot less state than a graphics pipeline:
```
```suggestion
As compute is not a part of the graphics pipeline, we can't use `vkCreateGraphicsPipelines`. Instead we need to create a dedicated compute pipeline with `vkCreateComputePipelines` for running our compute commands. Since a compute pipeline does not touch any of the rasterization state, it has a lot less state than a graphics pipeline:
```

SaschaWillems commented:
This is because our clone of the repo is behind the Vulkan Tutorial main repo. I just checked against that repo, and the PR will only include the compute changes, so we should be fine.

SaschaWillems committed:
Added links to advanced topics (tutorials, samples)

SaschaWillems committed:
Fixed typo

SaschaWillems committed:
Add note for the storage image chapter

SaschaWillems commented:
I did some small changes:
- As discussed on the last call, I added links to the advanced topics to some Khronos blogs and the samples
- Added a note to the storage image chapter that this is not part of the code that's demonstrated here, but rather just there to make people aware of compute and image manipulation
- Added note on how to handle uneven no. of invocations using gl_GlobalInvocationID

SaschaWillems committed:
Added note on how to handle uneven no. of invocations using gl_GlobalInvocationID

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+
+## Compute space
+
+Before we get into how a compute shader works and how we submit compute workloads to the GPU, we need to talk about two important compute concepts: **work groups** and **invocations**. They define an abstract execution model for how compute workloads are processed by the compute hardware of the GPU in three dimensions (x,y and z).
```
```suggestion
Before we get into how a compute shader works and how we submit compute workloads to the GPU, we need to talk about two important compute concepts: **work groups** and **invocations**. They define an abstract execution model for how compute workloads are processed by the compute hardware of the GPU in three dimensions (x, y, and z).
```
(oxford comma)

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+
+**Work groups** define how the compute workloads are formed and processed by the the compute hardware of the GPU. You can think of them as work items the GPU has to work through. Work group dimensions are set by the application at command buffer time using a dispatch command.
+
+And each work group then is a collection of **invocations** that execute the same compute shader. Invocations can potentially run in parallel and their dimensions are set in the compute shader. Invocations within a single workgroup have access to shared memory.
```
```suggestion
Each work group is a collection of **invocations** that all execute the same compute shader. Invocations can potentially run in parallel and their dimensions are set inside the compute shader. Invocations within a single workgroup have access to shared memory.
```
Shared memory is not something that has been defined above nor seems to have a definition below. Either there should be a clarification of what shared memory is here or removal of the sentence "Invocations ... have access to shared memory".

SaschaWillems replied:
That's right. Does this make more sense?

"One thing that's unique to compute is that invocations within a single workgroup can share variables using shared memory. This is an advanced concept not shown in this tutorial though."

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+
+This image shows the relation between these two in three dimensions:
+
+![](/images/compute_space.svg)
```
Maybe this image should have `Dispatch [3,3,3]`, indicating that its a 3x3x3 dispatch? What I'm thinking is that there is no obvious indication of why there is a 3 wide cube being shown.

Similar thought applies to the invocations.

SaschaWillems replied:
So something like this?

![image](https://user-images.githubusercontent.com/10283127/206767821-49c03b6b-8077-4501-807b-8fc365978ad3.png)

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+
+The number of dimensions for work groups and invocations depends on how input data is structured. If you e.g. work on a one-dimensional array, like we do in this chapter, you only have to specify the x dimension for both.
+
+As an example: If we dispatch a work group count of [64, 1, 1] with a compute shader local size of [32, 32, ,1], our compute shader will be invoked 64 x 32 x 32 = 65,536 times.
```
I don't think its obvious that `local size` refers to the invocations.

SaschaWillems replied:
I have clarified this a bit.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+
+As an example: If we dispatch a work group count of [64, 1, 1] with a compute shader local size of [32, 32, ,1], our compute shader will be invoked 64 x 32 x 32 = 65,536 times.
+
+Note that the max. count for work groups and local sizes differs from implementation to implementation, so you should always check the compute related limits in `VkPhysicalDeviceLimits`.
```
```suggestion
Note that the maximum count for work groups and local sizes differs from implementation to implementation, so you should always check the compute related limits in `VkPhysicalDeviceLimits`.
```
Maybe we should mention the exact limits that need to be used?  `maxComputeWorkGroupCount` `maxComputeWorkGroupInvocations` and `maxComputeWorkGroupSize` are the three related ones I see.

But this is your call, not a necessary change.

SaschaWillems replied:
That's actually a good point. Since I already mention `VkPhysicalDeviceLimits` it makes sense to put in the actual limit names.

charles-lunarg commented on en/11_Compute_Shader.md:
```diff
@@ -0,0 +1,605 @@
+std::vector<VkFence> computeInFlightFences;
+std::vector<VkSemaphore> computeFinishedSemaphores;
+...
+computeFinishedSemaphores.resize(MAX_FRAMES_IN_FLIGHT);
```
No resize for the computeInFlightFences?

SaschaWillems replied:
Good catch. The sample code actually does this, just forgot to add it to the tutorial.

charles-lunarg reviewed (commented):
Did a good pass on the grammar/typos as well as overall structure.

SaschaWillems committed:
Added compute fence resize to tutorial

Was already present in code

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems committed:
Update en/11_Compute_Shader.md

Co-authored-by: Charles Giessen <46324611+charles-lunarg@users.noreply.github.com>

SaschaWillems commented:
@charles-lunarg: Thank you very much for this thorough review. I merged all the spelling/grammar ones and will look at the remaining ones asap.

SaschaWillems committed:
Name the actual compute related physical device limits

SaschaWillems committed:
Clarfiy how work group and invocation counts are specified

SaschaWillems committed:
Show how initial particle data is generated for the shader storage buffer s

SaschaWillems committed:
Restructure SSBO setup

SaschaWillems committed:
Indentation

SaschaWillems committed:
Minor touches

SaschaWillems committed:
fixed transposition

SaschaWillems committed:
Added note on std140 memory layout qualifier

SaschaWillems commented:
The new chapter has been merged a few weeks ago, so closing this internal PR.
